### DISCUSSION TITLE
Message Passing on Undirected Bipartite Graphs

### DISCUSSION THREAD

**Discussion by Angelo92Git** [pyg-team/pytorch_geometric#9129] [Q&A]
Consider message passing on an undirected bipartite graph. Is there any gain in efficiency by setting the size parameter in the propagate function? How would the two sets of nodes be split up into sets of size (M, N)?

https://pytorch-geometric.readthedocs.io/en/latest/generated/torch_geometric.nn.conv.MessagePassing.html#torch_geometric.nn.conv.MessagePassing

**Comment by rusty1s** [comment]
For bi-directional message passing on bipartite graphs, I suggest to just utilize two GNN layers, i.e.:
```python
conv_A_B = SAGEConv((feat_A, feat_B), 32)
conv_B_A = SAGEConv((feat_B, feat_A), 32)

out_B = conv_A_B((in_A, in_B), edge_index_A_B)
out_A = conv_B_A((in_B, in_A), edge_index_B_A)
```
This is also what `HeteroConv` and `to_hetero` are doing internally.

**Reply by Angelo92Git** [reply]
Hi, just a follow-up question: How does the propagate function in the MessagePassing base class handle bipartite graphs?

1. Does this work if ```x``` is a 2-tuple, with the first entry being the features of nodes in the first set and the second entry containing the features of the second set?
2. Does this work if the nodes in the first and second set do not have the same dimensions
3. How does the size parameter affect this function, it seems that it is not a required parameter.

```
out = self.propagate(edge_index, size=size, x=x, norm=norm)
```

Thanks in advance!

**Reply by rusty1s** [reply]
> Does this work if x is a 2-tuple, with the first entry being the features of nodes in the first set and the second entry containing the features of the second set?

Yes.

> Does this work if the nodes in the first and second set do not have the same dimensions

Yes. In this case input features should be given as a tuple, e.g., `SAGEConv((16, 32), 64)`.

> How does the size parameter affect this function, it seems that it is not a required parameter.

It's not really needed. If you still want to pass it in, it should be `size=(x_src.size(0), x_dst.size(0))`.
